### PR TITLE
Refactor gate interface.

### PR DESCRIPTION
**Pull request by sergeisakov**

### CODE REVIEW

File: docs/type_reference.md [hunk 1]
```diff
@@ -19,7 +19,9 @@ they fulfill the same expectations.
 | Function (parfor.h)     | Any function; args are specified with `Args`.     |
 | FuserT                  | [`BasicGateFuser`](lib/fuser_basic.h)             |
 | Gate                    | [`Gate`](lib/gate.h)                              |
+| GateQSim                | [`GateQSim`](lib/gates_def.h)                     |
 | GateDef                 | [`GateRX`](lib/gatedef.h)                         |
+| GK / GateKind           | [`Gate`](lib/gate.h)
```
Comment (95-martin-orion) [inline docs/type_reference.md]: This link should probably be [`GateKind`](lib/gate.h) to distinguish from the `Gate` type above.

File: lib/gates_def.h [hunk 2]
```diff
@@ -26,6 +26,33 @@
...
+  kGateCZ,      // CZ
+  kGateCNot,    // CNOT (CX)
+  kGateIS,      // iSwap
+  kGateFS,      // fSim
+  kGateCP,      // control phase
+  kGateDecomp,  // single qubit gate from Schmidt decomposition
+};
+
+template <typename fp_type>
+using GateQSim = Gate<fp_type, GateKind>;
```
Comment (95-martin-orion) [inline lib/gates_def.h]: Personally, `GateQSim` doesn't help me distinguish between this and the base `Gate` type (aren't all gates qsim gates?) Maybe `DefaultGate` or similar would be clearer, what do you think?

File: docs/type_reference.md (hunk 1 above)
Comment (sergeisakov) [reply docs/type_reference.md]: Fixed.

File: lib/gates_def.h (hunk 2 above)
Comment (sergeisakov) [reply lib/gates_def.h]: Basically, GateQSim refers to the gates implemented in qsim. Gate is a generic gate. The idea is to make it easier to use qsim with external gate sets, which differ from the gate set implemented in qsim.

Comment (95-martin-orion) [reply lib/gates_def.h]: Okay, in the context of external gate sets that naming makes sense. To clarify this meaning for users, could you add a comment in the code here?

Comment (sergeisakov) [reply lib/gates_def.h]: Done.

**Pull request merged**